src/nsrt_learning.py

```diff
@@ -228,14 +227,168 @@ def learn_strips_operators(
+    """A skeleton is covered if all preconditions hold and final atoms
+    are predicted from the init atoms.
+    """
+    assert len(pnads) == len(orig_pnad_params)
+    name_to_idx = {pnad.op.name: idx for idx, pnad in enumerate(pnads)}
+    current_atoms = init_atoms
+    for (op_name, orig_objects) in skeleton:
+        idx = name_to_idx[op_name]
+        pnad = pnads[idx]
+        orig_params = orig_pnad_params[idx]
+        # Some parameters may have been removed, but order is preserved.
+        objects = tuple(o for o, v in zip(orig_objects, orig_params)
+                        if v in pnad.op.parameters)
+        ground_op = pnad.op.ground(objects)
+        if not ground_op.preconditions.issubset(current_atoms):
+            return False
+        current_atoms = utils.apply_operator(ground_op, current_atoms)
+    return final_atoms.issubset(current_atoms)
+
+
+def _recompute_datastores_from_segments(
+        segmented_trajs: List[List[Segment]],
+        pnads: List[PartialNSRTAndDatastore]) -> List[Set[Tuple[int, int]]]:
+    for pnad in pnads:
+        pnad.datastore = []  # reset all PNAD datastores
+    all_indices: List[Set[Tuple[int, int]]] = [set() for _ in pnads]
+    for traj_idx, seg_traj in enumerate(segmented_trajs):
+        objects = set(seg_traj[0].states[0])
+        for segment_idx, segment in enumerate(seg_traj):
+            identifier = (traj_idx, segment_idx)
+            (segment_param_option,
+             segment_option_objs) = segment.get_option_spec()
+            # Get ground operators given these objects and option objs.
+            for pnad_idx, pnad in enumerate(pnads):
+                param_opt, opt_vars = pnad.option_spec
+                if param_opt != segment_param_option:
+                    continue
+                isub = dict(zip(opt_vars, segment_option_objs))
+                # Consider adding this segment to each datastore.
+                for ground_op in utils.all_ground_operators_given_partial(
+                        pnad.op, objects, isub):
+                    # Check if preconditions hold.
+                    if not ground_op.preconditions.issubset(segment.init_atoms):
+                        continue
+                    # Check if effects match. Note that we're using the side
+                    # predicates semantics here!
+                    atoms = utils.apply_operator(ground_op, segment.init_atoms)
+                    if not atoms.issubset(segment.final_atoms):
+                        continue
+                    # This segment belongs in this datastore, so add it.
+                    # TODO: this is an edge case we need to fix by changing
+                    # ObjToVarSubs in the partition to VarToObjSubs.
+                    if len(set(ground_op.objects)) != len(ground_op.objects):
+                        continue
+                    sub = dict(zip(ground_op.objects, pnad.op.parameters))
+                    pnad.add_to_datastore((segment, sub))
```

Comment:
after fixing the bug below, this line will throw an error, because the assertions that we have in this method no longer make sense. you can just remove them or update them.